File: CPP/thread_example.cpp
// File: thread_example.cpp
// Created on: 23rd Dec 2018
// Updated on: 23rd Dec 2018
// Author: Pawan Kumar Singh [pawan@codingbull.com]
// http://www.codingbull.com
//
// This source code is the part of free tutorial series provided by http://www.codingbull.com and
// it can be used/distributed without any restriction.

#include <iostream>
#include <thread>

using namespace std;

void print_atoz() {
    for(char i = 'a'; i <= 'z'; i++)
        cout << i << endl;
}

struct thread_ex {
    void operator()(){
      print_1to100();
    }
    void print_1to100() {
      for(int i = 1; i <= 100; i++)
          cout << i << endl;
    }
};

int main() {
   cout << " ** Thread Example ** " << endl;
   thread_ex thread_example;
   //create new threads and execute them
   thread t1(thread_example);
   thread t2(print_atoz);

   //Wait untill both thread stops executing
   t1.join();
   t2.join();

   return 0;
}

File: Python/thread_example.py
# File: thread_example.py
# Created on: 23rd Dec 2018
# Updated on: 23rd Dec 2018
# Author: Pawan Kumar Singh [pawan@codingbull.com]
# http://www.codingbull.com
#
# This source code is the part of free tutorial series provided by http://www.codingbull.com and
# it can be used/distributed without any restriction.

import threading

def print_atoz():
    for i in range(ord('a'), ord('z')+1):
        print(chr(i))

def print_1to100():
    for i in range(1,100):
        print(i)

if __name__ == "__main__":
    # creating thread
    t1 = threading.Thread(target=print_atoz, args=())
    t2 = threading.Thread(target=print_1to100, args=())

    # starting threads
    t1.start()
    t2.start()

    # wait until threads are completely executed
    t1.join()
    t2.join()
